daemon/stats/collector_unix.go

```diff
@@ -54,15 +62,20 @@ func (s *Collector) getSystemCPUUsage() (uint64, error) {
 			for _, i := range parts[1:8] {
 				v, err := strconv.ParseUint(i, 10, 64)
 				if err != nil {
-					return 0, fmt.Errorf("Unable to convert value %s to int: %s", i, err)
+					return 0, fmt.Errorf("unable to convert value %s to int: %s", i, err)
 				}
 				totalClockTicks += v
 			}
-			return (totalClockTicks * nanoSecondsPerSecond) /
-				clockTicksPerSecond, nil
+
+			// This may result in a uint64 overflow which will result in an incorrect number
+			// being reported. However, since all CPU calculations are currently done
+			// in relation to one another, they will be correct after both points being
+			// compared have overflowed.
```

Comment:
I'm honestly a bit confused here -- isn't it possible that one call to `getSystemCPUUsage` doesn't overflow and then a subsequent call does, causing them to be a hugely disparate in the wrong direction?

(The way we calculate CPU usage percentage is comparing change in this value over time, right?)

Reply:
Everything you said is 100% correct. The key thing to note here is that once both values being compared both overflow, the difference between them will continue to remain the same.

Let's pretend our variable can only hold values from 0 to 100 and we collect two values from `getSystemCPUUsage`.

Val1 is 50
Val2 is 101

Because Val2 overflowed, Val2 is now stored as 0. 

Obviously, this calculation will yield the wrong value since Val2 - Val1 will be -50.

Now let's say we got another reading:

Val3 is 151

Now both of our values overflowed and the difference between them will be correct because Val2 will be 0 and Va3 will be 50. Therefore we have a difference of 50 and our calculations will be correct.

Here's a more concrete example: https://play.golang.org/p/FhdiJDntaxO

Reply:
See comments https://github.com/moby/moby/pull/41921#discussion_r564754647 and https://github.com/moby/moby/pull/41921#discussion_r564758518 for reference

Reply:
Ah, I had missed the "it will give out wrong information once" part, sorry!  That matches exactly what I'm concerned about :joy::innocent:

Reply:
No worries, thanks for the review! 😄